Finish cell typing text

jashapiro opened the pull request:
This PR hopefully completes the Cell Typing instruction notebook, closing #567.

I added a fair amount of text, but also made a few other modifications, like switching the reference data set to the one that celldex specifically recommends for PBMCs. This meant a few changes to the ways I used `forcats`... Nothing too major, but that chunk got a lot longer. 

Other than that, a few smaller changes, but nothing too major. I briefly contemplating using `scater::ggcells()` to allow for easier customization of the plots (in particular color schemes), but decided to leave it as is for now. That didn't seem worth getting too hung up about. If somebody asks, we might want to have that in our back pockets though!

I do update to using `|>` throughout as well, and tried a brief explanation of `\(x) x/sum(x)` but probably failed there. 

I have also designated as `live` the chunks I think I will want to type live (notably not the refactoring!). If there are others that seem worth enlivening, please suggest!

Otherwise, I look forward to hearing what you think

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -315,185 +316,324 @@ While this technique might be good for some simple experiments, and can be usefu
+Another component we would like to explore is how many of each of these cell types we have in the data set. 
+A bit of quick `dplyr` wrangling can give us the answer.
 
 ```{r}
````
chunk label

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -315,185 +316,324 @@ While this technique might be good for some simple experiments, and can be usefu
+### Running `SingleR`
+
+For our first run, we will do the marker gene selection (training) and classification in a single step, using the convenience function `SingleR::SingleR()`.
+For this we need only supply three main arguments: Our `sce` object, a reference matrix (here in `SummarizedExperiment` format), and the labels for each of the samples in the reference that we want to use.
```
I think we generally have been saying SCE (caps, not `code`), yes?

sjspielman reviewed (commented):
Left a couple small comments here before my brain totally crashes, but it's now crashing so we'll come back to this in 2023! 🎆

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -315,185 +316,324 @@ While this technique might be good for some simple experiments, and can be usefu
+The idea is that this set of genes will be the most informative for differentiating cell types.
 
-```{r}
+Then, for each cell, `SingleR` then calculates the Spearman correlation between expression that cell and each cell type (using the only the genes chosen earlier).
````
```suggestion
Then, for each cell, `SingleR` calculates the Spearman correlation between expression of that cell and each cell type (using the only the genes chosen earlier).
```

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -315,185 +316,324 @@ While this technique might be good for some simple experiments, and can be usefu
-```{r}
+Then, for each cell, `SingleR` then calculates the Spearman correlation between expression that cell and each cell type (using the only the genes chosen earlier).
+Notably, this is a non-parametric correlation, so the scaling and normalization that we apply (or don't) should not matter!
+(But if you have full-length transcripts from another technology though, you will probably want to convert to Transcripts per Million (TPM).)
````
```suggestion
(But if you have full-length transcripts from another technology, you will probably want to convert to Transcripts per Million (TPM).)
```

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -315,185 +316,324 @@ While this technique might be good for some simple experiments, and can be usefu
 
-If we were classifying multiple samples, we would want to do the training only once!
+We should note here that the reference dataset for `SingleR` does not need to be from a compendium like `celldex`!
+If you have any well-classified data set that you want to use as a reference, you can, as long as you can create a gene by sample expression matrix and a vector of cell types for each sample. 
```
```suggestion
If you have any well-classified dataset that you want to use as a reference, you can, as long as you can create a gene by sample expression matrix and a vector of cell types for each sample. 
```

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -315,185 +316,324 @@ While this technique might be good for some simple experiments, and can be usefu
-If we were classifying multiple samples, we would want to do the training only once!
+We should note here that the reference dataset for `SingleR` does not need to be from a compendium like `celldex`!
+If you have any well-classified data set that you want to use as a reference, you can, as long as you can create a gene by sample expression matrix and a vector of cell types for each sample. 
+You will want to ensure that the cell types you expect to see are present in the reference data set, and data should be normalized, but otherwise the method can be quite flexible. 
```
```suggestion
You will want to ensure that the cell types you expect to see in your sample are present in the reference dataset, and data should be normalized, but otherwise the method can be quite flexible. 
```
I'm changing to `dataset` throughout because in the beginning that's what you were using. Either one is fine, but I think we just want to be consistent.

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -315,185 +316,324 @@ While this technique might be good for some simple experiments, and can be usefu
 ```
 
-Now we plot.
+Now that we have that set up, we can now plot using our collapsed and ordered cell type labels.
````
```suggestion
Now that we have that set up, we can plot using our collapsed and ordered cell type labels.
```

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -315,185 +316,324 @@ While this technique might be good for some simple experiments, and can be usefu
-Now we plot.
+Now that we have that set up, we can now plot using our collapsed and ordered cell type labels.
 
 ```{r}
````
chunk label

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -315,185 +316,324 @@ While this technique might be good for some simple experiments, and can be usefu
 
 ### Heatmap of cell types & clusters
 
+Let's look at how the cell type assignments we have compare to the clusters that we found using the unsupervised clustering at the start of this notebook.
```
```suggestion
Let's look at how the cell type assignments we obtained using `SingleR` compare to the clusters that we found using the unsupervised clustering at the start of this notebook.
```

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -315,185 +316,324 @@ While this technique might be good for some simple experiments, and can be usefu
-To do this, we will first use `table` function to build a contingency table of the cell types and clusters that each cell was classified with. 
+To do this, we will again use the `table()` function, but now with two vectors as input, to build a contingency table of the cell types and clusters that each cell was classified with. 
 
 ```{r}
````
chunk label

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -504,89 +644,116 @@ There is a lot of customization we could do here, but `pheatmap` (pretty heatmap
 pheatmap::pheatmap(type_cluster_tab)
```
chunk label

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -504,89 +644,116 @@ There is a lot of customization we could do here, but `pheatmap` (pretty heatmap
+As there are only 20 clusters to classify, this will be very quick, and we don't need to parallelize it!
+
+```{r singler cluster, live = TRUE}
+singler_cluster <- SingleR::classifySingleR(
````
I think I would add in some comments here about each of the arguments.

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -504,89 +644,116 @@ There is a lot of customization we could do here, but `pheatmap` (pretty heatmap
 
-```{r}
-sce$celltype_cluster <- singler_result_cluster$labels[sce$nn_cluster]
+The result is a fairly small table of results, but we are interested most in the labels, which we would like to assign back to our SCE object cell by cell for visualization.
````
```suggestion
The result is a fairly small table of results, but we are most interested in the labels, which we would like to associate with each cell in our SCE object for visualization.
```

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -504,89 +644,116 @@ There is a lot of customization we could do here, but `pheatmap` (pretty heatmap
-Plot with cluster labels.
+Now we can plot these cluster-based cell type assignments using the now familiar `plotUMAP()` function
 
 ```{r}
````
chunk label

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -504,89 +644,116 @@ There is a lot of customization we could do here, but `pheatmap` (pretty heatmap
-Inspired by https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1812-2 
-(I may refine to match methods better, but maybe this is good enough?)
+As an intermediate step between the potentially messy single-cell cell type assignment and the almost certainly overconfident cluster-based assignment above, we can take an approach inspired by a paper by [Baran _et al._ (2019)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1812-2) that they called _metacells_. 
+The idea is that we can perform fine-scaled clustering to identify groups of similar very similar cells, then sum the counts within those clusters  for further analysis. 
```
```suggestion
The idea is that we can perform fine-scaled clustering to identify groups of similar cells, then sum the counts within those clusters for further analysis. 
```

allyhawkins commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -504,89 +644,116 @@ There is a lot of customization we could do here, but `pheatmap` (pretty heatmap
-We have ~8000 cells, so lets cluster those into groups of ~40 cells, which means we want to make about 200 clusters.
+To begin, we will perform some fine-scale clustering, using a simpler clustering algorithm: K-means clustering.
+We will use the same `bluster` package, clustering based on the PCA results we have from earlier, but this algorithm allows us to specify the number of clusters we want to end up with.
+We have  about 8000 cells, so let's cluster those into groups of approximately 80 cells, which works out to 100 clusters.
```
```suggestion
We have about 8000 cells, so let's cluster those into groups of approximately 80 cells, which works out to 100 clusters.
```

allyhawkins reviewed (commented):
This generally looks pretty good. I mostly had suggestions for cleaning up sentences and found some typos and missing chunk labels. The one thing I did notice throughout was the inconsistency between `dataset` and `data set`.

jashapiro commented:
Thanks for the reviews! I believe I have addressed the comments that were made here. I added missing chunk labels, standardized on "dataset" throughout, and fixed many of the other suggestions. 

Based on @sjspielman's initial comment on her earlier review, I would not be surprised if she had another round of comments before this is ready to go in. 

I realized that I also don't have anything about using one SCE to annotate another: I might add that in a separate PR if we think it is worth doing. Or it could perhaps be put in the exercise?

allyhawkins reviewed (approved):
This looks good to me!

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
 
-First, we need some reference data.
-Text about importance of reference data goes here!
+An alternative approach to using known marker genes for classification is to instead classify cells by way of comparison to a reference dataset.
```
```suggestion
An alternative approach to using known marker genes for classification is to instead classify cells by comparing to a reference expression dataset.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
-Text about importance of reference data goes here!
+An alternative approach to using known marker genes for classification is to instead classify cells by way of comparison to a reference dataset.
+To do this, we will find a well-curated gene expression dataset that contains samples with known cell types.
+We can then train a model based on this dataset and look at each of the cells in our new dataset to determine which (if any) of the known cell types has the most similar expression pattern.
```
```suggestion
We can then train a model based on this reference dataset and then ask whether any cells in our dataset have similar expression patterns.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+An alternative approach to using known marker genes for classification is to instead classify cells by way of comparison to a reference dataset.
+To do this, we will find a well-curated gene expression dataset that contains samples with known cell types.
+We can then train a model based on this dataset and look at each of the cells in our new dataset to determine which (if any) of the known cell types has the most similar expression pattern.
+The details of how such a model may be constructed and trained will vary by method, but this overall approach is widely applied. 
```
>will vary by method

We probably need to state explicitly that this method is more of the concept "cell annotation by reference" and not a specific method eg singler.

sjspielman replied:
Ah! Well then something definitely needs to be clarified since I got a bit lost in antecedents the first time around. Reading it again, I feel like something as simple as "...will vary by specific method" or "will vary by implementation" does it?

jashapiro replied:
```suggestion
The details of how such a model may be constructed and trained will vary by the specific method, but this overall approach is widely applied. 
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+We can then train a model based on this dataset and look at each of the cells in our new dataset to determine which (if any) of the known cell types has the most similar expression pattern.
+The details of how such a model may be constructed and trained will vary by method, but this overall approach is widely applied. 
+
+For this section, we will focus on the `SingleR` package and its methods, which are described in detail in  [_The SingleR Book_](https://bioconductor.org/books/3.16/SingleRBook/).
```
space 
```suggestion
For this section, we will focus on the `SingleR` package and its methods, which are described in detail in [_The SingleR Book_](https://bioconductor.org/books/3.16/SingleRBook/).
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+### Reference datasets
+
+Selecting a reference dataset is one of the more critical steps for this enterprise.
+At the most basic level, if the reference dataset does not include the types of cells that we expect to see in our sample, it won't be useful.
```
I don't think this is needed

```suggestion
If the reference dataset does not include the types of cells that we expect to see in our sample, it won't be useful.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+
+Selecting a reference dataset is one of the more critical steps for this enterprise.
+At the most basic level, if the reference dataset does not include the types of cells that we expect to see in our sample, it won't be useful.
+So we will want a dataset that has as many as possible of the cell types that we expect to see, at a level of specificity that aligns with our goals.
```
```suggestion
So we will want a reference dataset that has as many as possible of the cell types that we expect to are in our dataset, at a level of specificity that aligns with our goals.
```

specificity feels a little too stats-y for the concept? "granularity" maybe?

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+At the most basic level, if the reference dataset does not include the types of cells that we expect to see in our sample, it won't be useful.
+So we will want a dataset that has as many as possible of the cell types that we expect to see, at a level of specificity that aligns with our goals.
+
+For `SingleR` that reference data can be from bulk RNA sequencing or from other single-cell experiments.
```
```suggestion
`SingleR` analysis commonly uses bulk RNA sequencing or other single-cell experiments as references.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
 
-```{r}
-colData(blueprint_encode)
+```{r explore reference sample data}
````
live?

jashapiro replied:
```suggestion
```{r explore reference sample data, live=TRUE}
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+Another component we would like to explore is how many of each of these cell types we have in the dataset. 
+A bit of quick `dplyr` wrangling can give us the answer.
+
+```{r count cell types}
````
I think this could be live too, but no strong feelings.

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
 ```
 
+This is pretty good! 
+Most cell types have 4 replicates, which is more replicates than we often find.
````
Not sure about the word "replicate." 
"occurrence"? Or just something like, "we see most cell types at least 4x, which is good"

jashapiro replied:
Why not? These are biological replicates!

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+As mentioned earlier, `SingleR` builds a model from a set of training data, and then uses that model to classify cells (or groups of cells) in new datasets.
 
-This time we will do the training and classification in a single step, using the convenience function `SingleR::SingleR()`.
+`SingleR` works by first identifying a set of marker genes that can be used to differentiate among the cell types in the reference dataset. 
```
```suggestion
To build this model, `SingleR` first identifies a set of marker genes that can be used to differentiate among the cell types in the reference dataset. 
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
 
-This time we will do the training and classification in a single step, using the convenience function `SingleR::SingleR()`.
+`SingleR` works by first identifying a set of marker genes that can be used to differentiate among the cell types in the reference dataset. 
+It does this by performing pairwise comparisons among all of the cell types, and retaining the top set of genes differentiating each pair.
```
> performing pairwise comparisons

I think if you want to get into this you'll need a just a smidge more detail. What are the comparison tests and what is a "top" result?

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
-```{r}
+Then, for each cell, `SingleR` calculates the Spearman correlation between expression of that cell and each cell type (using the only the genes chosen earlier).
+Notably, this is a non-parametric correlation, so the scaling and normalization that we apply (or don't) should not matter!
+(But if you have full-length transcripts from another technology, you will probably want to convert to Transcripts per Million (TPM).)
````
```suggestion
[But if you have full-length transcripts from another technology, you will probably want to convert to Transcripts per Million (TPM).]
```

jashapiro replied:
Is this because there are nested parentheses? If it really bothers you, I would go the other way, with square internal parens. But I'll just unnest:
```suggestion
Note that if you used a single-cell technology that produces full-length transcripts (i.e., SMART-seq), you will probably want to convert your counts to Transcripts per Million (TPM), to allow more consistent ranking among transcripts of different lengths. 
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+### Running `SingleR`
+
+For our first run, we will do the marker gene selection (training) and classification in a single step, using the convenience function `SingleR::SingleR()`.
+For this we need only supply three main arguments: Our SCE object, a reference matrix (here in `SummarizedExperiment` format), and the labels for each of the samples in the reference that we want to use.
```
We might want to re-iterate here that's IDs do need to match, and they may recall we asked for Ensembl earlier so it would match our data.

jashapiro replied:
```suggestion
For this we need only supply three main arguments: Our SCE object, a reference matrix (here in `SummarizedExperiment` format), and the labels for each of the samples in the reference that we want to use.
We also need to be sure that our sample and the reference data use the same gene IDs, which is why we requested the Ensembl IDs when getting the reference dataset.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+```{r plot SingleR heatmap, live = TRUE}
 SingleR::plotScoreHeatmap(singler_result)
 ```
+We can also  pull out individual components of the results object for plotting in the context of our input SCE object.
````
space
```suggestion
We can also pull out individual components of the results object for plotting in the context of our input SCE object.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+```{r plot celltype umap, live=TRUE}
+scater::plotUMAP(sce, colour_by = "celltype_main") 
 ```
+Annoyingly, the `NA` and `T cells` labels are quite close in color, and the `scater` and `SingleR` packages don't agree on color choices, but this isn't a bad start!
````
😂 
```suggestion
Annoyingly, the `NA` and `T cells` labels are quite close in color, and the `scater` and `SingleR` packages don't agree on color choices, but this plot isn't a bad start!
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+scater::plotUMAP(sce, colour_by = "celltype_main") 
 ```
+Annoyingly, the `NA` and `T cells` labels are quite close in color, and the `scater` and `SingleR` packages don't agree on color choices, but this isn't a bad start!
+We seem to have a pretty good set of assignments, with most of the cell type assignments falling into groupings consistent with what we see in the UMAP plot.
````
```suggestion
We seem to have a pretty good set of cell type assignments, with most falling into groupings consistent with what we see in the UMAP plot.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+Quite often with other kinds of samples (especially cancer cells!) things will be much less clean!
 
-This is a pretty clean plot, for which we can thank that this is a PBMC data set.
+We can also look to see how the cell type assignments are distributed using the base R function `table()`
```
```suggestion
We can also look to see how the cell type assignments are distributed using the base R function `table()`.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+To do the training, we will use the `trainSingleR()` function. 
+For this we will start with our reference and the labels we want to train the model with.
+
+We can then specify the method used to select the genes that will be used for classification.
+The default method is `"de"`, which performs a differential expression analysis for each pair of labels, but we could also use `"sd"` to select the genes which are most variable across labels, or `"all"` to use all genes.
+If we want to get really fancy, we could even provide a specific list of genes to use.
```
Might link a singler book section here?

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+If you have any well-classified dataset that you want to use as a reference, you can, as long as you can create a gene by sample expression matrix and a vector of cell types for each sample. 
+You will want to ensure that the cell types you expect to see in your sample are present in the reference dataset, and data should be normalized, but otherwise the method can be quite flexible. 
+You can even use a previously-annotated SingleCellExperiment as a reference for a new dataset.
+For more details about custom references, see  the [OSCA chapter on cell type annotation](http://bioconductor.org/books/3.13/OSCA.basic/cell-type-annotation.html#using-custom-references)
```
space
```suggestion
For more details about custom references, see the [OSCA chapter on cell type annotation](http://bioconductor.org/books/3.13/OSCA.basic/cell-type-annotation.html#using-custom-references)
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
   sce,
+  # the trained model object
   singler_finemodel,
+  # perform fine tuning (default)
+  fine.tune = TRUE,
```
is this being written just to be explicit? fine!

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
-We'll do this by dividing each cluster/cell type count by the total number of cells in that cluster.
+As you can see, this produced a table with rows for each cell type and columns for each cluster number.
+The values are the count of cells for each cluster/cell type combination.
+These raw counts of cells are not quite what we are going to want for visualization. 
```
```suggestion
However, these raw counts are not quite what we'll want for visualization. 
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+As you can see, this produced a table with rows for each cell type and columns for each cluster number.
+The values are the count of cells for each cluster/cell type combination.
+These raw counts of cells are not quite what we are going to want for visualization. 
+Since the total number of cells in each cluster will vary, we'd like to convert these counts into the proportions of each cell type in each cluster.
```
```suggestion
Since the total number of cells differs across clusters, we'd like to convert these counts into the _proportions_ of each cell type in each cluster.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
 
+We'll do this by going through the table column by column and dividing each value by the sum for that cluster.
+To do that, we will use the `apply` function, which allows us to operate on a matrix row by row or column by column, applying a function to each "slice".
+Since the function we want to apply is very short, we will use R's new anonymous function shorthand: 
```
"new anonymous function shorthand" some kind of release link could be good? Mostly I mean, what is "NEW" in this case?

jashapiro replied:
How about just this:

```suggestion
Since the function we want to apply is very short, we will use R's new (as of  version 4.1) anonymous function shorthand: 
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
-`SingleR` is non-parametric, so it shouldn't matter! 
-(And adding log values felt bad.)
+The first step here is to create a new matrix where we sum the counts across cells that are from the same type according to our clustering.
+Because `SingleR` is a non-parametric approach, we will perform this step with the raw counts matrix.
```
```suggestion
Because `SingleR` is a non-parametric approach, we can perform this step with the raw counts matrix.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
-singler_result_cluster <- SingleR::classifySingleR(
-  sce_cluster_mat, 
-  singler_finemodel
+You can see that the resulting matrix still has the same number of rows we have seen before, but now only has rows for the number of clusters that were assigned.
```
```suggestion
You can see that the resulting matrix still has the same number of columns we have seen before, but now only has rows for the number of clusters that were assigned.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
 
-This sure looks nice, but what have we done here? 
-We are _assuming_ that each cluster has only a single cell type!
+This sure looks nice, but what have we really done here? 
```
```suggestion
This sure looks nice and clean, but what have we really done here? 
```

i like the clean word here, is mostly my point.

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+### MetaCell approaches
 
-### MetaCell-like
+As an intermediate step between the potentially messy single-cell cell type assignment and the almost certainly overconfident cluster-based assignment above, we can take an approach inspired by a paper by [Baran _et al._ (2019)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1812-2) that they called _metacells_. 
+The idea is that we can perform fine-scaled clustering to identify groups of very similar cells, then sum the counts within those clusters for further analysis. 
```
Wording is a little confusing here, reads like the paper is called _metacells_.
Open for back and forth here.

```suggestion
As a middle-ground between the potentially messy single-cell cell type assignment and the almost-certainly overconfident cluster-based assignment above, we can take approach inspired by [Baran _et al._ (2019)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1812-2) using something called _metacells_. 
The idea is that we can perform fine-scaled clustering to identify groups ("metacells") of very similar cells, then sum the counts within those clusters for further analysis. 
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
-### MetaCell-like
+As an intermediate step between the potentially messy single-cell cell type assignment and the almost certainly overconfident cluster-based assignment above, we can take an approach inspired by a paper by [Baran _et al._ (2019)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1812-2) that they called _metacells_. 
+The idea is that we can perform fine-scaled clustering to identify groups of very similar cells, then sum the counts within those clusters for further analysis. 
+The original paper includes a number of optimizations to make sure that the metacell clusters have desired properties for downstream analysis.
```
```suggestion
The original paper includes a number of optimizations to make sure that the metacell clusters have desirable properties for downstream analysis.
```

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+To save disk space (and time), we won't write out the whole SCE object, as we haven't changed any of the core data there. 
 Instead we will just write out the cell information table (`colData`) as a TSV file.
 
 ```{r save sce}
````
live?

jashapiro replied:
Needs a renaming too: 
````suggestion
```{r save cell info | live=TRUE}
````

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+We are _assuming_ that each cluster has only a single cell type, which is a pretty bold assumption, as we really aren't sure that the clusters we created were correct.
+You may recall that clustering algorithms are quite sensitive to parameter choice, so a different parameter choice could quite likely give a different result.
 
+### MetaCell approaches
```
I think this is good content, but just want to note we _could_ end up hitting fatigue/time limits for this section. That said the chunks are not live (good!), so this may end up serving as a useful reference for them. Will be interesting to see how it all goes!

sjspielman commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
```diff
@@ -309,284 +310,453 @@ Note that while we applied this technique to assign cell types using the ADT dat
+Is this more or less useful than the original cell-based clustering?
+
+
 ### Save results
```
level 2 here
```suggestion
## Save results
```

sjspielman reviewed (approved):
Looks good to go (in spite of my many comments)!! Don't need to see again unless you want me to look. I have a couple smaller comments for areas not explicitly in this PR but could be addressed now also -


2023!??
https://github.com/AlexsLemonade/training-modules/blob/8bd3eed8baff5c29eafefc254118c1b41a9d1a95/scRNA-seq-advanced/02-celltype_assignment.Rmd#L4

----------------

Noting we’ll want to circle back to this name
https://github.com/AlexsLemonade/training-modules/blob/8bd3eed8baff5c29eafefc254118c1b41a9d1a95/scRNA-seq-advanced/02-celltype_assignment.Rmd#L33

----------------

I might toss a new line in here
https://github.com/AlexsLemonade/training-modules/blob/8bd3eed8baff5c29eafefc254118c1b41a9d1a95/scRNA-seq-advanced/02-celltype_assignment.Rmd#L88

----------------

I think it would be good to have an explicit logic link here - surface proteins can be indicative of cell type so if something has a lot of cd3 it’s probably a t-cell.
https://github.com/AlexsLemonade/training-modules/blob/8bd3eed8baff5c29eafefc254118c1b41a9d1a95/scRNA-seq-advanced/02-celltype_assignment.Rmd#L175


----------------

I think we should quickly _remind_ here that this is an RNA UMAP colored by ADT, since I imagine this could be a point of confusion.
https://github.com/AlexsLemonade/training-modules/blob/8bd3eed8baff5c29eafefc254118c1b41a9d1a95/scRNA-seq-advanced/02-celltype_assignment.Rmd#L183

----------------

I’d add a quick “why” along with this. Maybe explain more up front - our goal is a long data frame with columns ADT and value for plotting purposes, so how do we get there? Follow along for a bunch of wrangling and explanations along the way.
https://github.com/AlexsLemonade/training-modules/blob/8bd3eed8baff5c29eafefc254118c1b41a9d1a95/scRNA-seq-advanced/02-celltype_assignment.Rmd#L223

----------------

Miscellaneously noting that `case_when()` is getting a bit more formally introduced in the integration notebook when we make the patient column.  
https://github.com/AlexsLemonade/training-modules/blob/8bd3eed8baff5c29eafefc254118c1b41a9d1a95/scRNA-seq-advanced/02-celltype_assignment.Rmd#L288

----------------


Do we ever actually say that this is a `colData` shortcut explicitly? Could be worth adding a comment?
https://github.com/AlexsLemonade/training-modules/blob/8bd3eed8baff5c29eafefc254118c1b41a9d1a95/scRNA-seq-advanced/02-celltype_assignment.Rmd#L301

jashapiro commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -136,7 +138,7 @@ Note that this PCA matrix and the UMAP built from it were derived from the gene
 
 
-```{r}
+```{r cluster cells}
````
I'll make it live... little warmup
```suggestion
```{r cluster cells, live=TRUE}
```

jashapiro commented on scRNA-seq-advanced/02-celltype_assignment.Rmd:
````diff
@@ -309,287 +310,457 @@ Note that while we applied this technique to assign cell types using the ADT dat
 
-```{r save sce}
+```{r save cell info, live=TRUE}
 readr::write_tsv(as.data.frame(colData(sce)), file = cellinfo_file)
````
Pipes for unnesting.
```suggestion
colData(sce) |>
  as.data.frame() |>
  readr::write_tsv(file = cellinfo_file)
```
